### PR TITLE
[Messenger][AMQP] Automatically reconnect on connection loss

### PR DESCRIPTION
**Pull request by ostrolucky**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      |no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

When using Rabbitmq in cluster, there is a common need of having to upgrade the nodes, while keeping the existing connections. The way this is normally done is by putting nodes in cluster to `maintenance mode`, ensuring cluster is healthy at all times. However, symfony/messenger nor php-amqp handle this use case at the moment. What happens instead is that exception when getting the message is thrown, worker crashes, error is logged and process manager has to respin it. This all happens without having a way in user space to handle this case better. Messenger's retry mechanism does not work here, because that one kicks in only when exception is thrown in handlers. Concrete exception is following:

>  [AMQPConnectionException (320)]
>  Server connection error: 320, message: CONNECTION_FORCED - Node was put into maintenance mode

What I'm proposing in this PR is that if connection error is detected _try to reconnect once_ before throwing exception. That should handle the outlined case. This goes line in line with recommendation from AWS's support we got:
> kindly ensure that the client connecting to the broker attempts a retry in case the above error message is observed during a maintenance window. In RabbitMQ cluster deployments, the nodes are restarted one-by-one, meaning at least two nodes will be up and running at all times. Even if a connection is severed, a connection retry will result in the other nodes accepting the connection, and the clients can keep using the broker.

I've also reported issue at https://github.com/php-amqplib/php-amqplib/issues/1161 with hope that this could be fixed at some point upstream, but I don't give it a big chance.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php [hunk 1]
```diff
@@ -52,6 +52,12 @@ private function getEnvelope(string $queueName): iterable
     {
         try {
             $amqpEnvelope = $this->connection->get($queueName);
+        } catch (\AMQPConnectionException) {
+            // Try to reconnect once to accommodate need for one of the nodes in cluster needing to stop serving the
+            // traffic. This may happen for example when one of the nodes in cluster is going into maintenance node.
+            // see https://github.com/php-amqplib/php-amqplib/issues/1161
+            $this->connection->queue($queueName)->getConnection()->reconnect();
+            $amqpEnvelope = $this->connection->get($queueName);
```
Comment (Nyholm) [inline src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php]: What if this line throws an exception? 
I think we must add try/catch around these two new lines too.

Comment (OskarStark) [reply src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php]: If I understand correct, here we want the exception as we only wants try to reconnect once, as mentioned in the description?

Comment (ostrolucky) [reply src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php]: Yeah if it fails second time, exception being thrown this time is fair

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php]: When it fails, we should convert the `AMQPException` exception to a `TransportException` one, right?

Comment (ostrolucky) [reply src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpReceiver.php]: Oh I see. Will do.

**Comment by ostrolucky** [comment]
ci failures unrelated

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ostrolucky.

**Pull request merged**